src/main/java/graphql/schema/validation/TypeAndFieldRule.java

```diff
@@ -0,0 +1,274 @@
+import graphql.schema.GraphQLEnumType;
+import graphql.schema.GraphQLScalarType;
+import graphql.schema.GraphQLEnumValueDefinition;
+import graphql.schema.GraphQLNonNull;
+import graphql.schema.GraphQLTypeUtil;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * The validation about GraphQLObjectType, GraphQLInterfaceType, GraphQLUnionType, GraphQLEnumType, GraphQLInputObjectType, GraphQLScalarType.
+ * <pre>
+ *     <ur>
+ *         <li>Types must define one or more fields;</li>
+ *         <li>Enum type must define one or more enum values;</li>
+ *         <li>Union type must include one or more unique member types;</li>
+ *         <li>The member types of a Union type must all be Object base types;</li>
+ *         <li>Non‐Null type must not wrap another Non‐Null type;</li>
+ *         <li>Invalid name begin with "__" (two underscores).</li>
+ *     </ur>
+ * </pre>
+ * <p></p>
+ * details in https://spec.graphql.org/June2018/#sec-Type-System
+ */
+public class TypeAndFieldRule implements SchemaValidationRule {
+
+    private Map<String, GraphQLNamedType> schemaTypeHolder;
+
+    @Override
+    public void check(GraphQLSchema graphQLSchema, SchemaValidationErrorCollector validationErrorCollector) {
+
+        List<GraphQLNamedType> allTypesAsList = graphQLSchema.getAllTypesAsList();
+
+        List<GraphQLNamedType> filteredType = filterScalarAndIntrospectionTypes(allTypesAsList);
+
+        schemaTypeHolder = graphQLSchema.getTypeMap();
+
+        checkTypes(filteredType, validationErrorCollector);
+    }
+
+
+    private void checkTypes(List<GraphQLNamedType> customizedType, SchemaValidationErrorCollector errorCollector) {
+        if (customizedType == null || customizedType.isEmpty()) {
+            return;
+        }
+
+        for (GraphQLType type : customizedType) {
+            checkType(type, errorCollector);
+        }
+    }
+
+    private void checkType(GraphQLType type, SchemaValidationErrorCollector errorCollector) {
+        if (type instanceof GraphQLObjectType) {
+            validateObject((GraphQLObjectType) type, errorCollector);
+        } else if (type instanceof GraphQLInterfaceType) {
+            validateInterface((GraphQLInterfaceType) type, errorCollector);
+        } else if (type instanceof GraphQLUnionType) {
+            validateUnion((GraphQLUnionType) type, errorCollector);
+        } else if (type instanceof GraphQLEnumType) {
+            validateEnum((GraphQLEnumType) type, errorCollector);
+        } else if (type instanceof GraphQLInputObjectType) {
+            validateInputObject((GraphQLInputObjectType) type, errorCollector);
+        } else if (type instanceof GraphQLScalarType) {
+            validateScalar((GraphQLScalarType) type, errorCollector);
+        }
+    }
+
+    private void validateObject(GraphQLObjectType type, SchemaValidationErrorCollector errorCollector) {
```

Comment:
can we combine validateObject and validateInterface?
I would also suggest a more prescriptive name like `validateContainsField` or so.

Reply:
Thanks for your suggestion.